src/test/scala/latis/util/hapi/HapiBinaryEncoderSpec.scala

```diff
@@ -0,0 +1,121 @@
+package latis.util.hapi
+
+import cats.effect.unsafe.implicits.global
+import org.scalatest.EitherValues._
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers._
+import scodec.bits._
+import scodec.codecs.implicits._
+import scodec.{Encoder => SEncoder, _}
+
+import latis.data.Data._
+import latis.data._
+import latis.dataset.Dataset
+import latis.dsl.DatasetGenerator
+import latis.metadata.Metadata
+import latis.model._
+import latis.util.Identifier.IdentifierStringContext
+
+class HapiBinaryEncoderSpec extends AnyFlatSpec {
+
+  import latis.output.BinaryEncoder
+
+  /** Instance of BinaryEncoder for testing. */
+  val enc: BinaryEncoder = HapiBinaryEncoder.hapiBinaryEncoder
+
+  "A Binary encoder" should "encode a dataset to binary" in {
+    val ds: Dataset = DatasetGenerator("x -> (a: int, b: double)")
+    val encodedList = enc.encode(ds).compile.toList.unsafeRunSync()
+
+    val expected = List(
+      SEncoder.encode(0).require.reverseByteOrder ++
+        SEncoder.encode(0).require.reverseByteOrder ++
+        SEncoder.encode(0.0).require.reverseByteOrder,
+      SEncoder.encode(1).require.reverseByteOrder ++
+        SEncoder.encode(1).require.reverseByteOrder ++
+        SEncoder.encode(1.0).require.reverseByteOrder,
+      SEncoder.encode(2).require.reverseByteOrder ++
+        SEncoder.encode(2).require.reverseByteOrder ++
+        SEncoder.encode(2.0).require.reverseByteOrder
+    )
+
+    encodedList should be(expected)
+  }
```

Comment:
I'm torn on whether to test this for every `Codec[Data]`.

Reply:
I think it might be good to have, just to make sure the new codec is actually getting used properly by the encoder.

Reply:
tl;dr: I don't think this is necessary in general but we can leave this for now since this is only the second and there are more important things to do.

My first line of thinking here is that correctly using the `Sample => Codec[Data]` argument is a concern of `BinaryEncoder`, not of whatever `Sample => Codec[Data]` we make.

My second line of thinking is that testing several `Sample => Codec[Data]` arguments stops telling us anything interesting after maybe two examples.

Assuming we follow the rules:
- `Sample => Codec[Data]` will always work
- The `BinaryEncoder` constructor can either:
    - Not use `Sample => Codec[Data]` and use some hard-coded `Codec[Data]`
    - Construct a `Codec[Data]` using the function, throw it away, and use some hard-coded `Codec[Data]`
    - Construct a `Codec[Data]` using the function but then apply some changes to it (limited by the algebras of `Codec` and `Data`)
    - Construct a `Codec[Data]` using the function and use it correctly

I think testing `BinaryEncoder` with a single `Sample => Codec[Data]` is enough to rule out the codec being modified. It could be the case that `BinaryEncoder` is using a hard-coded codec corresponding to the one our test provides, but testing with two `Sample => Codec[Data]` should rule that out. I'm not sure what a third example could rule out.

So, I don't think additional codecs past two add any additional assurances. This would be our second example, so maybe it could live here for now, but I think even adding a second `Codec[Data]` to the actual `BinaryEncoder` tests that does something as silly as turning each type into a constant would be sufficient as that second example. I think it's safe to just use a `Sample => Codec[Data]` where we've tested that the `Codec[Data]` works correctly for each type.

Note that this is not the same as testing with different shapes of datasets. That's like a separate dimension of the problem and I think testing different shapes with a single codec should be sufficient. Testing combinations of datasets and codecs seems like creating a MxN problem from an M+N problem.